ouroboros-consensus-cardano/src/cardano-testlib/Test/Consensus/Cardano/ProtocolInfo.hs

```diff
@@ -0,0 +1,361 @@
+                     allegraProtVer, alonzoMaxTxCapacityOverrides,
+                     alonzoProtVer, maryMaxTxCapacityOverrides, maryProtVer,
+                     sgGenDelegs, shelleyBasedGenesis, shelleyBasedInitialNonce,
+                     shelleyBasedLeaderCredentials,
+                     shelleyMaxTxCapacityOverrides, shelleyProtVer)
+import           Ouroboros.Consensus.Shelley.Node.Praos
+                     (ProtocolParamsBabbage (..), ProtocolParamsConway (..))
+import           Ouroboros.Consensus.Util.IOLike (IOLike)
+import qualified Test.ThreadNet.Infra.Alonzo as Alonzo
+import qualified Test.ThreadNet.Infra.Byron as Byron
+import qualified Test.ThreadNet.Infra.Shelley as Shelley
+import           Test.ThreadNet.Util.Seed (Seed (Seed), runGen)
+import           Test.Util.Slots (NumSlots (..))
+
+{-------------------------------------------------------------------------------
+  ProtocolInfo elaboration parameter types
+-------------------------------------------------------------------------------}
+
+newtype ByronSlotLengthInSeconds = ByronSlotLengthInSeconds Word64
+
+newtype ShelleySlotLengthInSeconds = ShelleySlotLengthInSeconds Word64
+
+class ToSlotLength a where
+  toSlotLength :: a -> SlotLength
+
+instance ToSlotLength ByronSlotLengthInSeconds where
+  toSlotLength (ByronSlotLengthInSeconds n) = Time.slotLengthFromSec $ fromIntegral n
+
+instance ToSlotLength ShelleySlotLengthInSeconds where
+  toSlotLength (ShelleySlotLengthInSeconds n) = Time.slotLengthFromSec $ fromIntegral n
+
+-- | This data structure is used to specify if and when hardforks should take
+-- place, and the version used at each era. See 'stayInByron' and 'hardForkInto'
+-- for examples.
+data HardForkSpec =
+    HardForkSpec {
+      shelleyHardForkSpec :: (SL.ProtVer, TriggerHardFork)
+    , allegraHardForkSpec :: (SL.ProtVer, TriggerHardFork)
+    , maryHardForkSpec    :: (SL.ProtVer, TriggerHardFork)
+    , alonzoHardForkSpec  :: (SL.ProtVer, TriggerHardFork)
+    , babbageHardForkSpec :: (SL.ProtVer, TriggerHardFork)
+    , conwayHardForkSpec  :: (SL.ProtVer, TriggerHardFork)
+    }
+
+data Era = Byron
+         | Shelley
+         | Allegra
+         | Mary
+         | Alonzo
+         | Babbage
+         | Conway
+  deriving (Show, Eq, Ord, Enum)
+
+selectEra :: Era -> HardForkSpec -> (SL.ProtVer, TriggerHardFork)
+selectEra Byron   _                                    = error "Byron is the first era, therefore there is no hard fork spec."
+selectEra Shelley HardForkSpec { shelleyHardForkSpec } = shelleyHardForkSpec
+selectEra Allegra HardForkSpec { allegraHardForkSpec } = allegraHardForkSpec
+selectEra Mary    HardForkSpec { maryHardForkSpec    } = maryHardForkSpec
+selectEra Alonzo  HardForkSpec { alonzoHardForkSpec  } = alonzoHardForkSpec
+selectEra Babbage HardForkSpec { babbageHardForkSpec } = babbageHardForkSpec
+selectEra Conway  HardForkSpec { conwayHardForkSpec  } = conwayHardForkSpec
```

Comment:
FTR: Functions like this could be simpler once we implement sth like https://github.com/input-output-hk/ouroboros-network/issues/4115

Reply:
I've added a comment to the function